neo/tools/imgui/particleeditor/ParticleEditor.cpp

```diff
@@ -0,0 +1,1585 @@
+			const idDict *dict = gameEdit->EntityGetSpawnArgs( list[i] );
+			if ( dict == NULL ) {
+				continue;
+			}
+			const char *name = dict->GetString( "name" );
+			gameEdit->EntitySetModel( list[i], val );
+			gameEdit->EntityUpdateVisuals( list[i] );
+			gameEdit->EntityGetAxis( list[i], axis );
+			//vectorControl.SetidAxis( axis );
+			gameEdit->MapSetEntityKeyVal( name, "model", val );
+		}
+		buttonSaveParticleEntitiesEnabled = true;
+	}
+}
+
+void ParticleEditor::AddStage( bool clone ) {
+
+	idDeclParticle *idp = GetCurParticle();
+	if ( idp == NULL ) {
+		return;
+	}
+
+	idParticleStage *stage = new idParticleStage;
+
+	if ( clone ) {
+		idParticleStage *source = GetCurStage();
+		if ( source == NULL ) {
+			delete stage;
+			return;
+		}
+		*stage = *source;
+	} else {
+		stage->Default();
+	}
+	int newIndex = idp->stages.Append( stage );
+	int index = listStages.Append( idStr::Format( "stage %i", newIndex ) );
+	listStagesSel = index;
+	listStagesItemData.Add( index, newIndex );
+	ShowCurrentStage();
+	EnableStageControls();
+}
+
+void ParticleEditor::RemoveStage() {
+	idDeclParticle *idp = GetCurParticle();
+	if ( idp == NULL ) {
+		return;
+	}
+
+	if ( !ImGui::InputMessageBox( "Are you sure you want to remove this stage?", "Remove Stage", true ) ) {
+		return;
+	}
+
+	int index = listStagesSel;
+	if ( index >= 0 ) {
+		int newIndex = listStagesItemData.First( index );
+		if ( newIndex >= 0 && newIndex < idp->stages.Num() ) {
+			idp->stages.RemoveIndex( newIndex );
+			index += ( index >= 1 ) ? -1 : 1;
+			newIndex = comboParticle.FindIndex( idp->GetName() );
+			EnumParticles();
+			if ( newIndex >= 0 ) {
+				comboParticleSel = newIndex;
+			}
+			OnCbnSelchangeComboParticles();
+			listStagesSel = index;
+			ShowCurrentStage();
+		}
+	}
+	EnableStageControls();
+}
+
+void ParticleEditor::ShowStage() {
+	idParticleStage *ps = GetCurStage();
+	if ( ps == NULL ) {
+		return;
+	}
+	ps->hidden = false;
+	int index = listStagesSel;
+	int newIndex = listStagesItemData.First( index );
+	listStages.RemoveIndex( index );
+	listStagesItemData.Remove( index, newIndex );
+	listStages.Insert( idStr::Format("stage %i", index ), index );
+	listStagesItemData.Add( index, newIndex );
+	listStagesSel = index;
+	EnableStageControls();
+}
+
+void ParticleEditor::HideStage() {
+	idParticleStage *ps = GetCurStage();
+	if ( ps == NULL ) {
+		return;
+	}
+	ps->hidden = true;
+	int index = listStagesSel;
+	int newIndex = listStagesItemData.First( index );
+	listStages.RemoveIndex( index );
+	listStagesItemData.Remove( index, newIndex );
+	listStages.Insert( idStr::Format("stage %i (H) ", index ), index );
+	listStagesItemData.Add( index, newIndex );
+	listStagesSel = index;
+	EnableStageControls();
+}
+
+idParticleStage *ParticleEditor::GetCurStage() {
+	idDeclParticle *idp = GetCurParticle();
+	int sel = listStagesSel;
+	int index = listStagesItemData.First( sel );
+	if ( idp == NULL || sel == LB_ERR || index >= idp->stages.Num() ) {
```

Comment:
LB_ERR doesn't make sense here, it's something Windows-specific (and makes the build fail on linux and other operating systems)

Reply:
Good catch. Fixed.